extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,151 @@
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(e, url, selector) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+    // Scan the page to get a list of all image ads on this page, indexed by url.
+    // For performance reasons, we only do the scan once, the first time that a batch manager is constructed.
+    e.win.imageadElements = {};
+    const elements = document.querySelectorAll(selector);
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const url = el.getAttribute('data-url');
+      if (!(url in e.win.imageadElements)) {
+        e.win.imageadElements[url] = [];
+      }
+      e.win.imageadElements[url].push(el);
+    }
+  }
+  if (!(url in e.win.imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    e.isBatchMaster_ = true;
+    e.win.imageadBatchManagers[url] = new AmpAdBatchManager(e);
+  } else {
+    e.isBatchMaster_ = false;
+  }
+  return e.win.imageadBatchManagers[url];
+}
+
+export class AmpAdBatchManager {
+
+  /** @param {!Element} element */
+  constructor(element) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = element.url_;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullUrl_ = this.url_;
+
+    /** @private {number} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.retries_ = AD_BATCH_RETRIES;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Get a list of slots that are in use for this URL
+    const elements = element.win.imageadElements[this.url_];
+    const slots = elements.map(element => {
+      const slotId = element.getAttribute('data-slot');
+      // Use a slot id of 0 if the element didn't provide one.
+      return slotId ? slotId : 0;
+    });
+    // If any slots were defined, add them to the URL
+    if (slots.length) {
+      this.fullUrl_ += '?s=' + slots.join(',');
+    }
+  }
+
+  /**
+   * Call this to instruct the batch manager to do the layout for an element. Prior to doing the layout, the batch manager
+   * will fetch the data (if this is the batch master) or wait for it to be ready (if not the batch master).
+   * @param {Element} elem The element for which the layout is to be done
+   * @returns {Promise} The element's "jsondata_" variable will be filled its with the data from the ad server before this is
+   *     resolved. If resolved OK, the element is returned by the promise
+   */
+  doLayout(elem) {
+    const t = this;
```

Comment:
Got it. That also explains why you prefer using arrow functions, which I hadn't previously understood.